### PR TITLE
destination-async-framework: move the state emission logic into GlobalAsyncStateManager

### PR DESCRIPTION
**Pull request by subodh1810**
There is no issue for this. See slack [thread](https://airbytehq-team.slack.com/archives/C0624HL9UTU/p1707843108366429?thread_ts=1707513683.088909&cid=C0624HL9UTU)

### CODE REVIEW

File: airbyte-cdk/java/airbyte-cdk/core/src/main/java/io/airbyte/cdk/integrations/destination_async/buffers/MemoryAwareMessageBatch.java [hunk 1]
```diff
@@ -64,9 +63,9 @@ public void close() throws Exception {
    *
    * @return list of states that can be flushed
```
Comment (davinchia) [inline airbyte-cdk/java/airbyte-cdk/core/src/main/java/io/airbyte/cdk/integrations/destination_async/buffers/MemoryAwareMessageBatch.java:64]: nit: change java doc.

File: airbyte-cdk/java/airbyte-cdk/core/src/main/java/io/airbyte/cdk/integrations/destination_async/buffers/MemoryAwareMessageBatch.java [hunk 2]
```diff
@@ -64,9 +63,9 @@ public void close() throws Exception {
...
    */
-  public List<PartialStateWithDestinationStats> flushStates(final Map<Long, Long> stateIdToCount) {
+  public void flushStates(final Map<Long, Long> stateIdToCount) {
     stateIdToCount.forEach(stateManager::decrement);
-    return stateManager.flushStates();
+    stateManager.flushStates();
```
Comment (davinchia) [inline airbyte-cdk/java/airbyte-cdk/core/src/main/java/io/airbyte/cdk/integrations/destination_async/buffers/MemoryAwareMessageBatch.java:68]: Curious, does the locking break if we return the ids to flush and flush them in the FlushWorkers instead?

That way, all the actual action happens in the flush workers, and not in objects scattered around. Mainly a cleanliness suggestion.

File: airbyte-cdk/java/airbyte-cdk/core/src/main/java/io/airbyte/cdk/integrations/destination_async/state/GlobalAsyncStateManager.java [hunk 3]
```diff
@@ -161,8 +164,7 @@ public void decrement(final long stateId, final long count) {
    *
    * @return list of state messages with no more inflight records.
    */
-  public List<PartialStateWithDestinationStats> flushStates() {
-    final List<PartialStateWithDestinationStats> output = new ArrayList<>();
+  public void flushStates() {
```
Comment (davinchia) [inline airbyte-cdk/java/airbyte-cdk/core/src/main/java/io/airbyte/cdk/integrations/destination_async/state/GlobalAsyncStateManager.java:167]: suggestion: what if `flushStates` took in an output record collector? This helps us prevent passing the record collector everywhere. Should help comprehension.

File: airbyte-cdk/java/airbyte-cdk/core/src/main/java/io/airbyte/cdk/integrations/destination_async/state/GlobalAsyncStateManager.java [hunk 4]
```diff
@@ -161,8 +164,7 @@ public void decrement(final long stateId, final long count) {
    *
    * @return list of state messages with no more inflight records.
```
Comment (davinchia) [inline airbyte-cdk/java/airbyte-cdk/core/src/main/java/io/airbyte/cdk/integrations/destination_async/state/GlobalAsyncStateManager.java:165]: nit: update javadoc.

**Review by davinchia — APPROVED** [review]
Conditional approval.

If I understand correctly, what we do here is centralise flushing into one class. Flushing itself is locked, so though many threads can call it, we prevent a race condition, which is good! Approved if my understanding is correct.

Most of my feedback is for readability. All action happens in the FlushWorkers in the previous iteration. This is intentional, since someone can read one class and understand most of what is going on. Signalling to another class is a fine pattern - my main suggestion is to keep the output record collector in one spot so it's clearer where writing mostly happens. This is just a suggestion.

What is a must-have, though non-blocking, is to update the javadocs. Keeping them in sync will really help readability!

File: airbyte-cdk/java/airbyte-cdk/core/src/main/java/io/airbyte/cdk/integrations/destination_async/state/GlobalAsyncStateManager.java [hunk 5]
```diff
@@ -95,11 +96,13 @@ public class GlobalAsyncStateManager {
   private long retroactiveGlobalStateId = 0;
   // All access to this field MUST be guarded by a synchronized(lock) block
   private long arrivalNumber = 0;
+  private final Consumer<AirbyteMessage> outputRecordCollector;
 
-  private final Object LOCK = new Object();
+  private static final Object LOCK = new Object();
```
Comment (gisripa) [inline airbyte-cdk/java/airbyte-cdk/core/src/main/java/io/airbyte/cdk/integrations/destination_async/state/GlobalAsyncStateManager.java:101]: static seems unnecessary ?

File: airbyte-cdk/java/airbyte-cdk/core/src/main/java/io/airbyte/cdk/integrations/destination_async/state/GlobalAsyncStateManager.java [hunk 6]
```diff
@@ -95,11 +96,13 @@ public class GlobalAsyncStateManager {
...
 
-  public GlobalAsyncStateManager(final GlobalMemoryManager memoryManager) {
+  public GlobalAsyncStateManager(final GlobalMemoryManager memoryManager, final Consumer<AirbyteMessage> outputRecordCollector) {
     this.memoryManager = memoryManager;
+    this.outputRecordCollector = outputRecordCollector;
```
Comment (gisripa) [inline airbyte-cdk/java/airbyte-cdk/core/src/main/java/io/airbyte/cdk/integrations/destination_async/state/GlobalAsyncStateManager.java:105]: Can we avoid holding a copy of outputRecordCollector at instance level and use only in flushStates ? This reduces the possibility of future abuse of this instance variable.

**Review by gisripa — CHANGES_REQUESTED** [review]

**Review by gisripa — APPROVED** [review]

**Pull request merged**